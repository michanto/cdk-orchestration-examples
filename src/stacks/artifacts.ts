import { AppToken, BUILD_TIME, Log, StackToken } from '@michanto/cdk-orchestration';
import { S3FileResource } from '@michanto/cdk-orchestration/orchestration';
import { App, Stack, StackProps } from 'aws-cdk-lib';
import { Function, Runtime, Code } from 'aws-cdk-lib/aws-lambda';
import { IBucket, Bucket } from 'aws-cdk-lib/aws-s3';
import { md5hash } from 'aws-cdk-lib/core/lib/helpers-internal';
import { PhysicalResourceId } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

const NAMESPACE = '@michanto/cdk-orchestration-examples';

/** Token names for use with AppToken and StackToken . */
const NAMED_BUCKET_TKN = `${NAMESPACE}.NamedBucket`;
const UNNAMED_BUCKET_TKN = `${NAMESPACE}.UnnamedBucket`;
const BUILD_ID_TKN = `${NAMESPACE}.BuildId`;

/**
 * Class that creates the tokens we use in our App.
 */
export class SampleTokens {
  /** Resolves to the name of the Named Bucket. */
  static namedBucketNameToken(scope: Construct) {
    return AppToken.string(scope, NAMED_BUCKET_TKN, { displayHint: 'NmdBkt' });
  }

  /**
   * Separate token creation from token resolution.
   * Useful for cross-stack references without having to pass constructs.
   */
  static resolveNamedBucketNameToken(scope: Construct) {
    AppToken.resolveString(scope, NAMED_BUCKET_TKN, {
      produce: () => {
        let stack = Stack.of(scope);
        return `artifact-bucket-${stack.account}-${stack.region}-${
          md5hash(stack.node.path).slice(0, 8).toLowerCase()
        }`;
      },
    });
  }

  /** Resolves to the name of the unnamed Bucket (name will be generated at deployment time). */
  static unnamedBucketNameToken(scope: Construct) {
    return AppToken.string(scope, UNNAMED_BUCKET_TKN, { displayHint: 'UnnmdBkt' });
  }

  /*
  * Separate token creation from token resolution.
  * Useful for cross-stack references without having to pass constructs.
  */
  static resolveUnnamedBucketNameToken(scope: Construct, bucket: IBucket) {
    AppToken.resolveString(scope, UNNAMED_BUCKET_TKN, {
      produce: () => {
        return bucket.bucketName;
      },
    });
  }

  /**
   * Token for the buildId, which defaults to a timestamp, but is configurable.
   *
   * At Synthesis time, this token is resolved by checking a dictionary attached to the
   * App construct.
   *
   * @param scope Any scope in the app will do, as this is an AppToken.
   * @returns A string that prints as `${Token[BLDID.<token id>]}`
   */
  static buildIdToken(scope: Construct) {
    return AppToken.string(scope, BUILD_ID_TKN, {
      displayHint: 'BLDID',
    });
  }

  /**
   * Resolves the token named "CdkOrche::BuildId"
   * Resolved to any, but CDK will call toString() when writing the value since the token
   * is a string. Note BUILD_TIME is a number.
   * @param scope Any scope in the app will do, as this is an AppToken.
   * @param value
   */
  static resolveBuildIdToken(scope: Construct, value: any = BUILD_TIME) {
    AppToken.resolveAny(scope, BUILD_ID_TKN, { produce: () => value });
  }
}


export class OtherStack extends Stack {
  constructor(scope: Construct, id: string = 'ArtifactBucketsStack', props?: StackProps) {
    super(scope, id, props);
    let log = Log.of(this);

    log.info('// @michanto/cdk-orchestration AppToken //');
    let buildIdToken = AppToken.string(this, 'build_id', {
      displayHint: 'BLDID',
    });
    log.info(JSON.stringify(buildIdToken));
    log.info(this.toJsonString(buildIdToken));

    // Application scoped named token.
    let bucketToken = AppToken.string(this, 'code_bucket', {
      displayHint: 'Bkt',
    });

    log.info('// @michanto/cdk-orchestration StackToken //');
    let app = this.node.root as App;

    let valinor = new Stack(app, `${this.stackName}-Valinor`, {
      env: {
        account: this.account, region: 'us-east-1',
      },
    });

    let beleriand = new Stack(app, `${this.stackName}-Beleriand`, {
      env: {
        account: this.account, region: 'eu-west-1',
      },
    });
    log.info('// Create Stack token in two different stacks. //');
    for (let stack of [valinor, beleriand]) {
      let keyToken = StackToken.string(stack, 'code_key', {
        displayHint: 'Key',
      });
      let codeBucket = Bucket.fromBucketName(stack, 'CodeBucket', bucketToken);
      new Function(stack, 'MyFunction', {
        runtime: Runtime.NODEJS_16_X,
        code: Code.fromBucket(codeBucket, keyToken),
        handler: 'handler',
      });
      StackToken.resolveString(stack, 'code_key', {
        produce: () => `${
          stack.stackName.toLowerCase()
        }/${
          buildIdToken
        }/code.zip`,
      });
    }
    let unresolvedToken = StackToken.string(this, 'unresolved_token', {
      displayHint: 'Unres',
    });
    log.info(unresolvedToken);
    try {
      log.info(`UnresolvedToken = '${this.resolve(unresolvedToken)}'`);
    } catch {
      log.info(`Could not resolve ${unresolvedToken}`);
    }

    StackToken.resolveString(this, 'unresolved_token', { produce: () => 'Now Resolved' });
    try {
      log.info(`UnresolvedToken = '${this.resolve(unresolvedToken)}'`);
    } catch {
      log.info(`Could not resolve ${unresolvedToken}`);
    }

    // Resolve app tokens.
    AppToken.resolveAny(this, 'build_id', { produce: () => BUILD_TIME });
    // Explicitly named bucket for use cross-region.
    let codeBucketName = `code-bucket-${this.account}`;
    new Bucket(this, 'CodeBucket', {
      bucketName: codeBucketName,
    });
    // Resolve bucket name.
    AppToken.resolveString(this, 'code_bucket', { produce: () => codeBucketName });

    log.info(this.resolve(buildIdToken));
  }
}
/**
 * The artifact bucket will be used to store artifact creation products.
 */
export class ArtifactBucketsStack extends Stack {
  constructor(scope: Construct, id: string = 'ArtifactBucketsStack', props?: StackProps) {
    super(scope, id, props);

    new Bucket(this, 'NamedBucket', {
      bucketName: SampleTokens.namedBucketNameToken(this),
    });
    SampleTokens.resolveNamedBucketNameToken(this);

    const unnamedBucket = new Bucket(this, 'UnnamedBucket', {
    });
    SampleTokens.resolveUnnamedBucketNameToken(this, unnamedBucket);
  }
}

/**
 * Rather than pass the ArtifactBucket to this class, it can use the token.
 */
export class ArtifactWriterStack extends Stack {
  constructor(scope: Construct, id: string = 'ArtifactWriter', props?: StackProps) {
    super(scope, id, props);
    let namedBucket = Bucket.fromBucketName(this, 'NamedBucket', SampleTokens.namedBucketNameToken(this));
    let unnamedBucket = Bucket.fromBucketName(this, 'UnnamedBucket', SampleTokens.unnamedBucketNameToken(this));

    new S3FileResource(this, 'Written2NamedBucket', {
      bucket: namedBucket,
      key: 'buildid.json',
      body: {
        buildId: SampleTokens.buildIdToken(this),
      },
      physicalResourceId: PhysicalResourceId.of('buildid.json'),
    });
    new S3FileResource(this, 'Written2UnnamedBucket', {
      bucket: unnamedBucket,
      key: 'buildid.json',
      body: {
        buildId: SampleTokens.buildIdToken(this),
      },
      physicalResourceId: PhysicalResourceId.of('buildid.json'),
    });
  }
}
