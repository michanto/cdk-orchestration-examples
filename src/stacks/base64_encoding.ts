import { Logger, LogLevel } from '@michanto/cdk-orchestration';
import { CustomResourceUtilities } from '@michanto/cdk-orchestration/custom-resources';
import { Echo, ImportOrders } from '@michanto/cdk-orchestration/transforms';
import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { GreetingLambdaTask } from '../constructs/greeting_lambda_task';

/**
 * Echo is a Transform, and Transforms work out-of-the-box with L1 or L2 constructs.
 * Otherwise, the default is for a transform to be applied to the stack.
 *
 * Since we want this Transform to be applied to the custom resource in an L3 construct
 * (such as Task or AwsCustomResource), we need to override the target property to
 * return the correct L1 construct.
 */
export class EchoCustomResource extends Echo {
  get target() {
    return new CustomResourceUtilities().findCustomResource(this.node.scope!);
  }
}

export class EchoCustomResourceAfterEncode extends EchoCustomResource {
  get order() {
    // Encode runs in WRITER order.
    return ImportOrders.WRITER;
  }
}

export class Base64EncodingExample extends Stack {
  constructor(scope: Construct, id: string = 'TransformsInfo', props?: StackProps) {
    super(scope, id, props);
    console.log('OrderedTransforms constructor BEGIN');
    Logger.set(this, new Logger({ logLevel: LogLevel.DEBUG }));

    /**
     * Similar to CDK Triggers, LambdaTask classes are custom resources that call a lambda.
     */
    let task = new GreetingLambdaTask(this, 'GreetingLambdaTask', {
      greeting: 'Hello, everyone!',
    });
    new EchoCustomResource(task, 'EchoCustomResource');
    new EchoCustomResourceAfterEncode(task, 'EchoCustomResourceAfterEncode');

    console.log('OrderedTransforms constructor END');
  }
}
