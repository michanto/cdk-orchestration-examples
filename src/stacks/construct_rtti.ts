import { ConstructRunTimeTypeInfo, ConstructTreeSearch, IStopCondition, Log, Logger, LogLevel } from '@michanto/cdk-orchestration';
import { App, CfnElement, Stack, StackProps } from 'aws-cdk-lib';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct, IConstruct } from 'constructs';
import { NAMESPACE } from '../private/internal';

/** A construct with RTTI for testing. */
export class TypedConstruct extends Construct {
  /** Type assertion for TypedConstruct. */
  static isTypedConstruct(x: IConstruct): x is TypedConstruct {
    return TypedConstruct.TYPED_CONSTRUCT_RTTI.hasRtti(x);
  }

  /**
   * Return the TypedConstruct hosting the given construct.
   */
  static of(x: IConstruct): TypedConstruct | undefined {
    let host = TypedConstruct.TYPED_CONSTRUCT_RTTI.searchUp(x);
    return host as TypedConstruct | undefined;
  }

  /**
   * Return all typed constructs under the given scope.
   */
  static typedConstructs(scope: IConstruct, stopCondition?: IStopCondition) {
    return ConstructTreeSearch.for(TypedConstruct.isTypedConstruct)
      .searchDown(scope, stopCondition);
  }

  private static readonly TYPED_CONSTRUCT_RTTI = new ConstructRunTimeTypeInfo({
    servicePropertyName: `${NAMESPACE}.test.TypedConstruct`,
  });

  constructor(scope: Construct, id: string) {
    super(scope, id);
    TypedConstruct.TYPED_CONSTRUCT_RTTI.addRtti(this);
  }
}

/**
 * Demonstrate built in and custom RTTI in the CDK.
 */
export class ConstructRtti extends Stack {
  constructor(scope: IConstruct, id: string, props?: StackProps) {
    super(scope, id, props);
    let app = this.node.root as App;
    Logger.set(app, new Logger({ logLevel: LogLevel.INFO }));
    let log = Log.of(this);

    // Built in RTTI
    log.info(`This construct is a stack: ${Stack.isStack(this)}`);
    log.info(`This construct is a construct: ${Construct.isConstruct(this)}`);
    log.info(`This construct is a CfnElement?: ${CfnElement.isCfnElement(this)}`);

    new TypedConstruct(this, 'UnderStack');
    new TypedConstruct(this.node.root, 'UnderApp');
    let bucket = new Bucket(this, 'MyBucket');
    new TypedConstruct(bucket, 'UnderBucket');
    new TypedConstruct(bucket.node.defaultChild!, 'UnderCfnBucket');
    let parent = new TypedConstruct(this, 'AParent');
    let hostedBucket = new Bucket(parent, 'HostedBucket');
    let subStack = new Stack(this, 'SubStack');
    new TypedConstruct(subStack, 'UnderSubStack');

    log.info(`isTypedConstruct true example: ${
      TypedConstruct.isTypedConstruct(parent)}`);
    log.info(`isTypedConstruct false example: ${
      TypedConstruct.isTypedConstruct(bucket)}`);
    log.info(`TypedConstructs in this app:\n${
      TypedConstruct.typedConstructs(app).map(c => c.node.path).join('\n  ')}`);
    log.info(`TypedConstructs in this stack:\n${
      TypedConstruct.typedConstructs(this, Stack.isStack).map(c => c.node.path).join('\n  ')}`);
    log.info(`TypedConstructs under this stack:\n${
      TypedConstruct.typedConstructs(this).map(c => c.node.path).join('\n  ')}`);
    log.info(`TypedConstructs in the sub stack:\n${
      TypedConstruct.typedConstructs(subStack).map(c => c.node.path).join('\n  ')}`);
    log.info(`TypedConstruct above hostedBucket:\n${
      TypedConstruct.of(hostedBucket)!.node.path}`);
    // Get the stack of all TypedConstructs under the Stack
    TypedConstruct.typedConstructs(this).forEach(t => Stack.of(t));
  }
}
