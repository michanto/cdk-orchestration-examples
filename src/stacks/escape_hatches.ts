import { BUILD_TIME, CfnElementUtilities, ConstructRunTimeTypeInfo, Log } from '@michanto/cdk-orchestration';
import { InlineNodejsFunction } from '@michanto/cdk-orchestration/aws-lambda-nodejs';
import { CustomResourceUtilities, RunResourceAlways } from '@michanto/cdk-orchestration/custom-resources';
import { CfnElement, CfnResource, Stack, StackProps } from 'aws-cdk-lib';
import { CfnRole } from 'aws-cdk-lib/aws-iam';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from 'aws-cdk-lib/custom-resources';
import { Trigger } from 'aws-cdk-lib/triggers';
import { Construct, IConstruct } from 'constructs';
import { NAMESPACE } from '../private/internal';

const LAMBDA_PATH = `${__dirname}/../../lib/constructs/lambdas/`;

export class StackDescription extends CfnElement {
  constructor(scope: Construct, id: string, readonly description: string) {
    super(scope, id);
  }

  _toCloudFormation(): object {
    return {
      Description: this.description,
    };
  }
}

/**
 * A well designed construct has a Properties interface with readonly members.
 */
export interface AddMetadataProperties {
  /** Metadata key. */
  readonly key: string;
  /** Metadata value. */
  readonly value: any;
}

/**
 * This construct adds metadata to a Stack or a CfnResource.
 *
 * A well designed construct has documentation.
 *
 * An well designed escape hatch construct works with Stacks (if applicable),
 * L1, L2 or L3 constructs without the user having to navigate
 * the construct tree.
 */
export class AddMetadata extends Construct {
  /** RTTI is optional depending on your use case. */
  static isAddMetadata(x: IConstruct): x is AddMetadata {
    return AddMetadata.ADD_METADATA_CONSTRUCT_RTTI.hasRtti(x);
  }

  private static readonly ADD_METADATA_CONSTRUCT_RTTI = new ConstructRunTimeTypeInfo({
    servicePropertyName: `${NAMESPACE}.AddMetadata`,
  });

  constructor(scope: Construct, id: string, readonly props: AddMetadataProperties) {
    super(scope, id);

    AddMetadata.ADD_METADATA_CONSTRUCT_RTTI.addRtti(this);
    this.target.addMetadata(props.key, props.value);
  }

  /**
   * By puttng the target in a getter, subclasses can easily override the target.
   * This expands the usefulness of our Construct.
   * For example, the user could target a specfic resource type in a sub tree.
   */
  get target(): CfnResource | Stack {
    let scope = this.node.scope!;

    // If this construct is added to a stack, the user wants to add stack metadata.
    // Otherwise, the user wants to add CfnResource metadata.
    if (Stack.isStack(scope)) {
      return scope;
    }

    // Otherwise the user wants to add metadata to a CfnResource under scope.
    return new CfnElementUtilities().findCfnResource(scope);
  }
}

/**
 * Adds Salt property to any custom resource.
 * NOTE:  This exists in cdk-orchestration as RunResourceAlways construct.
 *
 * In the CDK Triggers [documentation](
 * https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.triggers-readme.html) it says:
 * >> In the future we will consider adding support for additional re-execution modes:
 *
 * >> `executeOnEveryDeployment`: boolean - re-executes every time the stack is
 * deployed (add random "salt" during synthesis).
 *
 * We can go ahead and implement that, but for ANY custom resource.
 */
export class AddSalt extends Construct {
  constructor(scope: Construct, id: string = 'AddSalt', readonly saltValue: number = BUILD_TIME ) {
    super(scope, id);
    this.target.addPropertyOverride('salt', saltValue);
    Log.of(this).debug(() => `Added salt value ${saltValue}.`);
  }

  get target() {
    let scope = this.node.scope!;
    return new CustomResourceUtilities().findCustomResource(scope);
  }
}

/**
 * Show how to use [Escape Hatches](https://docs.aws.amazon.com/cdk/v2/guide/cfn_layer.html)
 * in the CDK.
 *
 * The 'cfn layer' document recommends using L1 methods and raw overrides as escape hatches.
 * Here, we'll see that we can encapsulate these as constructs and use searches to make working
 * with escape hatches more natural.
 */
export class EscapeHatches extends Stack {
  constructor(scope: Construct, id: string = 'WritingConstructs', props?: StackProps) {
    super(scope, id, { ...props, description: 'EscapeHatches description.' });
    let log = Log.of(this);

    new StackDescription(this, 'Description', 'This string will be in the stack description');
    new StackDescription(this, 'Description2', 'And so will this one.');

    new AddMetadata(this, 'StackMetadata', {
      key: 'Elf', value: 'Arondir',
    });

    let bucket = new Bucket(this, 'MyBucket');
    new AddMetadata(bucket, 'BucketMetadata', {
      key: 'Mountain', value: 'Taniquetil',
    });

    const echoLambda = new InlineNodejsFunction(this, 'EchoLambda', {
      entry: `${LAMBDA_PATH}/echo.js`,
      environment: {
        LogLevel: '1',
      },
    });

    new AddMetadata(echoLambda, 'FunctionMetadata', {
      key: 'Dwarf', value: 'Balin',
    });

    new class AddRoleMetadata extends AddMetadata {
      get target(): CfnResource | Stack {
        return new CfnElementUtilities().findCfnResource(this.node.scope!, CfnRole.CFN_RESOURCE_TYPE_NAME);
      }
    }(echoLambda, 'RoleMetadata', {
      key: 'Maia', value: 'Bombadillo',
    });

    let trigger = new Trigger(this, 'EchoTrigger', {
      handler: echoLambda,
    });
    log.debug(() => 'Echo lambda will trigger on every deployment.');
    new AddSalt(trigger);

    let writer = new AwsCustomResource(this, 'MyS3FileResource', {
      onUpdate: {
        service: 'S3',
        action: 'putObject',
        parameters: {
          Body: JSON.stringify({ dummy: 'data' }),
          Bucket: bucket.bucketName,
          Key: 'dummy.json',
        },
        physicalResourceId: PhysicalResourceId.of(`s3:://${bucket.bucketName}/dummy.json`),
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: [bucket.bucketArn, `${bucket.bucketArn}/*`],
      }),
    });
    new RunResourceAlways(writer);
  }
}
