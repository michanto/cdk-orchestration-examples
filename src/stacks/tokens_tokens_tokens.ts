import { readFileSync } from 'fs';
import { Log, Logger, StackUtilities } from '@michanto/cdk-orchestration';
import { App, Aws, Fn, IStableStringProducer, Lazy, Stack, StackProps, Token } from 'aws-cdk-lib';
import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

/**
 * In this lesson we learn:
 * - How to name buckets using tokens.
 * - How buckets use tokens to reference the bucket name.
 * - How to write a string token that resolves to CloudFormation.
 * - The difference between stable tokens and uncached tokens.
 * - How to write an inline lambda.
 * - Why you shouldn't call resolve on tokens except as a debugging aid.
 */
export class TokensTokensTokens extends Stack {
  constructor(scope: Construct, id: string = 'TokensTokensTokens', props?: StackProps) {
    super(scope, id, props);
    Logger.set(this, new Logger());

    let log = Log.of(this);
    log.info('//** Tokens, Tokens, Tokens! **//');

    log.info('// Bucket names and tokens');

    /**
     * One option: don't name the bucket.
     * S3 will create a name for you.
     */
    let unnamedBkt = new Bucket(this, 'UnnamedBucket', {
      bucketName: undefined,
    });
    /**
     * Note that the bucket.bucketName is a always a token that
     * resolved to a Fn::Ref for the S3 bucket, which always resolves
     * to the bucket name.
     */
    log.info(`UnnamedBucket name ref: '${unnamedBkt.bucketName}' resolved: ${
      !Token.isUnresolved(unnamedBkt.bucketName)} value: ${
      JSON.stringify(this.resolve(unnamedBkt.bucketName))
    }`);

    /** Buckets can also be named with constant string. */
    let staticName = 'static-named-000000000000-us-west-2';
    let staticNamedBucket = new Bucket(this, 'StaticNamedBucket', {
      bucketName: staticName,
    });
    log.info(`StaticNamedBucket name ref: '${staticNamedBucket.bucketName}' resolved: ${
      !Token.isUnresolved(staticNamedBucket.bucketName)} value: ${
      JSON.stringify(this.resolve(staticNamedBucket.bucketName))
    }`);
    log.info(`StaticNamedBucket name: ${staticName} resolved: ${
      !Token.isUnresolved(staticName)}`);

    /** Buckets can be named with an Fn.join */
    let joinName = Fn.join('-', ['join', 'named', Aws.ACCOUNT_ID, Aws.REGION]);
    new Bucket(this, 'JoinNamedBucket', {
      bucketName: joinName,
    });
    log.info(`JoinNamedBucket name: '${joinName}' resolved: ${
      !Token.isUnresolved(joinName)}`);

    /**
     * How does that work?  The Fn.join function returns a string,
     * but produces an ANY in CloudFormation?  It returns an Any token,
     * which can resolve to a number, string, or JSON, then calls Token.asString
     * to create a string token that resolves to the Any.
     * Note the desplayHint is passed to both the Any token and the string token.
     */
    let myJoin = function (delim: string, values: string[]) {
      return Token.asString(Lazy.any({
        produce: () => {
          // If all values are fully resolved, return the constant.
          if (values.every(val => !Token.isUnresolved(val))) {
            return values.join(delim);
          }
          return { 'Fn::Join': [delim, values] };
        },
      }, { displayHint: 'MyJoin' }), { displayHint: 'MyJoin' });
    };

    let allStringsResolved = myJoin('-', ['all', 'strings', 'resolved']);
    log.info('// My version of Join.');
    log.info(allStringsResolved);
    log.info(this.resolve(allStringsResolved));
    log.info(JSON.stringify(this.resolve(
      myJoin('-', ['some', 'strings', 'resolved', Aws.ACCOUNT_ID]))));

    let implicitJoinName = `implicit-join-name-${Aws.ACCOUNT_ID}-${Aws.REGION}`;
    new Bucket(this, 'ImplicitJoinNamedBucket', {
      bucketName: implicitJoinName,
    });
    log.info(`ImplicitJoinNamedBucket name: '${implicitJoinName}' resolved: ${
      !Token.isUnresolved(implicitJoinName)}`);

    let app = this.node.root as App;

    // Stable tokens
    log.info('// Stable tokens //');
    let spider = Lazy.string({ produce: () => 'Shelob' });
    log.info(spider);

    log.info('// Stable token with display hint');
    let spiderMother = Lazy.string({ produce: () => 'Ungoliant' },
      { displayHint: 'Spider' });
    log.info(spiderMother);

    // Resolve tokens
    log.info('// Resolve stable tokens');
    log.info(this.resolve(spider));
    log.info(this.resolve(spiderMother));

    // Uncached tokens
    log.info('// Uncached tokens //');

    let regionToken = Lazy.uncachedString({
      produce: (ctx) => Stack.of(ctx.scope).region,
    }, { displayHint: 'Region' });

    let valinor = new Stack(app, `${this.stackName}-Valinor`, {
      env: {
        account: this.account, region: 'us-east-1',
      },
    });

    let beleriand = new Stack(app, `${this.stackName}-Beleriand`, {
      env: {
        account: this.account, region: 'eu-west-1',
      },
    });
    log.info('// Print uncached token');
    log.info(regionToken);

    log.info('// Resolve in Valinor stack');
    log.info(valinor.resolve(regionToken));
    log.info('// Resolve in Beleriand stack');
    log.info(beleriand.resolve(regionToken));

    log.info('// Intrinsic token //');
    log.info(Aws.REGION);
    log.info(JSON.stringify(valinor.resolve(Aws.REGION)));
    log.info(JSON.stringify(beleriand.resolve(Aws.REGION)));

    log.info('// Complex intrinsic token //');
    let complexToken = Fn.getAzs(Aws.REGION);
    log.info(JSON.stringify(complexToken));
    log.info(JSON.stringify(valinor.resolve(complexToken)));
    log.info(JSON.stringify(beleriand.resolve(complexToken)));

    let allStacksProducer: IStableStringProducer = {
      produce: () => new StackUtilities().stacks(app).map(s => s.stackName).join(','),
    };

    /** By resolving this token now, it won't pick up stacks created later. */
    let someStacks = Lazy.string(allStacksProducer, { displayHint: 'Stacks' });
    this.resolve(someStacks);

    new Function(this, 'StackNamesEnv', {
      code: Code.fromInline(readFileSync(`${__dirname}/../../lib/constructs/lambdas/echo.js`).toString()),
      handler: 'index.handler',
      runtime: Runtime.NODEJS_20_X,
      environment: {
        STACK_NAMES: Lazy.string(allStacksProducer, { displayHint: 'Stacks' }),
        SOME_STACK_NAMES: someStacks,
        SPIDER: spider,
        SPIDER_MOTHER: spiderMother,
        UNNAMED_BUCKET: unnamedBkt.bucketName,
        STATIC_NAMED_BUCKET: staticNamedBucket.bucketName,
        REGION_TOKEN: regionToken,
      },
    });

    new Stack(app, `${this.stackName}-AnotherStack`);
  }
}