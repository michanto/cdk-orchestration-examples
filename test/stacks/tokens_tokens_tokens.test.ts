import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { TokensTokensTokens } from '../../src/stacks/tokens_tokens_tokens';

test('TokensTokensTokens', () => {
  const app = new App();
  const stack = new TokensTokensTokens(app);

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();

});
