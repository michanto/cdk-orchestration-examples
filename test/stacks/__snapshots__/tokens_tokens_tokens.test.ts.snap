// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokensTokensTokens 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ImplicitJoinNamedBucketDA919B54": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": {
          "Fn::Join": [
            "",
            [
              "implicit-join-name-",
              {
                "Ref": "AWS::AccountId",
              },
              "-",
              {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "JoinNamedBucket829C7202": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": {
          "Fn::Join": [
            "-",
            [
              "join-named",
              {
                "Ref": "AWS::AccountId",
              },
              {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "StackNamesEnv9BC0B335": {
      "DependsOn": [
        "StackNamesEnvServiceRole79CE8D59",
      ],
      "Properties": {
        "Code": {
          "ZipFile": ""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = handler;
function log(message) {
    if (process.env.LogLevel) {
        console.log(JSON.stringify(message));
    }
}
async function handler(event, context) {
    log({ Event: event });
    log({ Context: context });
    return Promise.resolve(event !== null && event !== void 0 ? event : {});
}
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiZWNoby5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uLy4uLy4uL3NyYy9jb25zdHJ1Y3RzL2xhbWJkYXMvZWNoby50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOztBQUtBLDBCQUlDO0FBVEQsU0FBUyxHQUFHLENBQUMsT0FBNEI7SUFDdkMsSUFBSSxPQUFPLENBQUMsR0FBRyxDQUFDLFFBQVEsRUFBRSxDQUFDO1FBQ3pCLE9BQU8sQ0FBQyxHQUFHLENBQUMsSUFBSSxDQUFDLFNBQVMsQ0FBQyxPQUFPLENBQUMsQ0FBQyxDQUFDO0lBQ3ZDLENBQUM7QUFDSCxDQUFDO0FBQ00sS0FBSyxVQUFVLE9BQU8sQ0FBQyxLQUFVLEVBQUUsT0FBWTtJQUNwRCxHQUFHLENBQUMsRUFBRSxLQUFLLEVBQUUsS0FBSyxFQUFFLENBQUMsQ0FBQztJQUN0QixHQUFHLENBQUMsRUFBRSxPQUFPLEVBQUUsT0FBTyxFQUFFLENBQUMsQ0FBQztJQUMxQixPQUFPLE9BQU8sQ0FBQyxPQUFPLENBQUMsS0FBSyxhQUFMLEtBQUssY0FBTCxLQUFLLEdBQUksRUFBRSxDQUFDLENBQUM7QUFDdEMsQ0FBQyIsInNvdXJjZXNDb250ZW50IjpbImZ1bmN0aW9uIGxvZyhtZXNzYWdlOiBSZWNvcmQ8c3RyaW5nLCBhbnk+KSB7XG4gIGlmIChwcm9jZXNzLmVudi5Mb2dMZXZlbCkge1xuICAgIGNvbnNvbGUubG9nKEpTT04uc3RyaW5naWZ5KG1lc3NhZ2UpKTtcbiAgfVxufVxuZXhwb3J0IGFzeW5jIGZ1bmN0aW9uIGhhbmRsZXIoZXZlbnQ6IGFueSwgY29udGV4dDogYW55KSB7XG4gIGxvZyh7IEV2ZW50OiBldmVudCB9KTtcbiAgbG9nKHsgQ29udGV4dDogY29udGV4dCB9KTtcbiAgcmV0dXJuIFByb21pc2UucmVzb2x2ZShldmVudCA/PyB7fSk7XG59XG4iXX0=",
        },
        "Environment": {
          "Variables": {
            "REGION_TOKEN": {
              "Ref": "AWS::Region",
            },
            "SOME_STACK_NAMES": "TokensTokensTokens,TokensTokensTokens-Valinor,TokensTokensTokens-Beleriand",
            "SPIDER": "Shelob",
            "SPIDER_MOTHER": "Ungoliant",
            "STACK_NAMES": "TokensTokensTokens,TokensTokensTokens-Valinor,TokensTokensTokens-Beleriand,TokensTokensTokens-AnotherStack",
            "STATIC_NAMED_BUCKET": {
              "Ref": "StaticNamedBucketF597C035",
            },
            "UNNAMED_BUCKET": {
              "Ref": "UnnamedBucketC3A6BB1E",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "StackNamesEnvServiceRole79CE8D59",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "StackNamesEnvServiceRole79CE8D59": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "StaticNamedBucketF597C035": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "static-named-000000000000-us-west-2",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "UnnamedBucketC3A6BB1E": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
